### PR TITLE
feat(nuxt): experimental flag to use external vue dependencies

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

https://github.com/nuxt/framework/issues/4084

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR starts exploring implementing moving `vue`, `@vue/*` and `vue-router` to externals. As it is, it is not the solution. Although it works, we have (at least) the following issues:

- [ ] multiple copies of imports are included, for almost all `vue` and `@vue` libs, meaning an increase of ~2.2M (688kB -> 2.9M)  in `node_modules` for a minimal project using `pages/` - but note that the **total** `.output/server` size _decreases_ slightly from 3.3M -> 3.2M. (**Note**: we can improve this into a much larger decrease by implementing import conditions upstream in vue.)
- [ ] there are extra, non-treeshaken bits of vue-router that we could prune (such as other forms of router history and all code behind `isBrowser` flag)
- [ ] unneeded _parts_ of the vue api, such as `@vue/devtools-api`, are included in server node_modules

Some ideas to explore:

* 'light' processing of Vue externals - extra compilation steps to strip browser or dev-only code?
* manual post-build step to remove dev cjs builds
* import conditions (development/production) for Vue core libraries

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by huang-julien** [comment]
HI @danielroe. What if we modify the externals plugin from nitro to avoid sending some files to the final bundle ? Would it work ?  We would be able to exclude the @vue/devtools-api from the server node_modules

by changing [this line](https://github.com/unjs/nitro/blob/a6aaafa57f8e0369cc09a91b5c1532c2ce570738/src/rollup/plugins/externals.ts#L149) to : 
```ts
then((r) => r.filter((file) => file.includes("node_modules") && !excludeExternals.some(name => file.includes(name)));
```


image

File: packages/nuxt/src/core/nitro.ts [hunk 1]
```diff
@@ -58,16 +58,25 @@ export async function initNitro (nuxt: Nuxt) {
     sourcemap: nuxt.options.sourcemap,
     externals: {
       inline: [
-        ...(nuxt.options.dev ? [] : ['vue', '@vue/', '@nuxt/', nuxt.options.buildDir]),
+        ...(nuxt.options.dev
+          ? []
+          : [
+              ...nuxt.options.experimental.externalVue ? [] : ['vue', '@vue/'],
+              '@nuxt/', nuxt.options.buildDir
```
Comment (pi0) [inline packages/nuxt/src/core/nitro.ts:65]:
```suggestion
              '@nuxt/',
              nuxt.options.buildDir
```

File: packages/schema/src/config/experimental.ts [hunk 2]
```diff
@@ -17,6 +17,12 @@ export default {
...
      * @see https://vuejs.org/guide/extras/reactivity-transform.html
      */
-    reactivityTransform: false
+    reactivityTransform: false,
+
+    /**
+     * Externalize `vue`, `@vue/*` and `vue-router` when build
+     * @see https://github.com/nuxt/framework/issues/4084
+     */
+    externalVue: false
```
Comment (pi0) [inline packages/schema/src/config/experimental.ts:26]: What do you think to generalize flag to as `externalDependencies`? So we have the chance to progressively extend it?

Comment (pi0) [reply packages/schema/src/config/experimental.ts:26]: It is short-termed but saw `vue-router` which is not vue exactly. Just taught of using a more generic flag.

Comment (danielroe) [reply packages/schema/src/config/experimental.ts:26]: I'd definitely be up for improving the name. But I guess it's about core vue libraries, and possibly `externalDependencies` is too broad. Can you think of a narrower name?

Comment (pi0) [reply packages/schema/src/config/experimental.ts:26]: External libs/modules might respect same experiment ad well. When vue is externalized, they need to also avoid inline their vue libs right? BTW it doesn't matter that much for naming I'm fine with current one too if you prefer to keep

File: packages/vite/src/server.ts [hunk 3]
```diff
@@ -30,16 +30,21 @@ export async function buildServer (ctx: ViteBuildContext) {
...
+          : {
+              '@vue/reactivity': _resolve(`@vue/reactivity/dist/reactivity.cjs${ctx.nuxt.options.dev ? '' : '.prod'}.js`),
+              '@vue/shared': _resolve(`@vue/shared/dist/shared.cjs${ctx.nuxt.options.dev ? '' : '.prod'}.js`),
+              'vue-router': _resolve(`vue-router/dist/vue-router.cjs${ctx.nuxt.options.dev ? '' : '.prod'}.js`)
+            },
         vue: _resolve(`vue/dist/vue.cjs${ctx.nuxt.options.dev ? '' : '.prod'}.js`)
       }
     },
     ssr: {
+      external: ctx.nuxt.options.experimental.externalVue ? ['#internal/nitro', 'vue', 'vue-router'] : [],
```
Comment (pi0) [inline packages/vite/src/server.ts:47]: Why `#internal/nitro` is in the condition? 🤔

Comment (danielroe) [reply packages/vite/src/server.ts:47]: It was there previously.

Comment (danielroe) [inline packages/vite/src/server.ts:47]:
```suggestion
      external: ctx.nuxt.options.experimental.externalVue ? ['#internal/nitro', 'vue', 'vue-router'] : ['#internal/nitro'],
```

**Review by pi0 — APPROVED** [review]

**Pull request merged**